```{r knitrSetup, include=FALSE}
# sets output for inline results nicely
#knit_hooks$set(inline = identity) 

# kills knitr if there is an error, therefore we don't waste time generating error messages
knit_hooks$set(error = function(x, options) stop(x)) 
```

```{r figTableTrack, include=FALSE}
# this function helps us insert figure and table text and links
pasteLabel <- function(preText, inObj, objName, insLink=TRUE){
	objNum <- inObj[objName]
	
	useText <- paste(preText, objNum, sep=" ")
	if (insLink){
		useText <- paste("[", useText, "](#", objName, ")", sep="")
	}
	useText
}

# this increments the counter, and gives a name to the number so we can reference it later
incCount <- function(inObj, useName){
	nObj <- length(inObj)
	useNum <- max(inObj) + 1
	inObj <- c(inObj, useNum)
	names(inObj)[nObj+1] <- useName
	inObj
}
figCount <- c("_"=0)
tableCount <- c("_"=0)
```

```{r otherFunctions, include=FALSE}
tableCat <- function(inFrame){
	outText <- paste(names(inFrame), collapse=" | ")
	outText <- c(outText, paste(rep("---", ncol(inFrame)), collapse=" | "))
	invisible(apply(inFrame, 1, function(inRow){
		outText <<- c(outText, paste(inRow, collapse=" | "))
	}))
	return(outText)
}

probeClass <- function(probeIDs){
	sapply(strsplit(probeIDs, ".", fixed=T), function(y){y[1]})
}

probeSetName <- function(probeID){
	splitName <- strsplit(probeID, ".", fixed=T)[[1]]
	nSplit <- length(splitName)
	rmSplit <- -1 * c(1, nSplit-1, nSplit)
	paste(splitName[rmSplit], sep=".", collapse=".")
}

sortProbes <- function(probeIDs){
	splitIDs <- strsplit(probeIDs, ".", fixed=T)
	endIndx <- sapply(splitIDs, function(x){
		nEnt <- length(x)
		as.numeric(x[c(nEnt-1, nEnt)])
	})
	endIndx <- t(endIndx)
	order(endIndx[,1], endIndx[,2])
}

```

```{r loadPackages, include=FALSE}
require(ggplot2)
require(plyr)
require(VennDiagram)
require(hgu133plus2.db)
require(hgu133plus2probe)
require(org.Hs.eg.db)

load("counts_Plots_etc_allOrgs.RData")
```

# Affymetrix(r) Mismatch (MM) Probes: Useful After All

Robert M Flight^(1), Abdallah M Eteleeb^(2) and Eric C Rouchka^(2)

^(1) Department of Anatomical Sciences and Neurobiology, University of Louisville, Louisville, Kentucky, USA 40292

^(2) Department of Computer Engineering and Computer Science, University of Louisville, Louisville, Kentucky, USA 40292

### Email Contacts

robert dot flight near louisville dot edu

ametel01 near louisville dot edu

eric dot rouchka near louisville dot edu


## ABSTRACT

Affymetrix(r) GeneChip(r) microarray design define probe sets consisting of 11, 16, or 20 distinct 25 base pair (BP) probes
for determining mRNA expression for a specific gene, which may be covered by one or more probe sets.  Each probe has a
corresponding perfect match (PM) and mismatch (MM) set.  Traditional analytical techniques have either used the MM probes
to determine the level of cross-hybridization or reliability of the PM probe, or have been completely ignored.  Given the
availability of reference genome sequences, we have reanalyzed the mapping of both PM and MM probes to reference genomes
in transcript regions.  Our results suggest that depending of the species of interest, 66%-93% of the PM probes can be 
used reliably in terms of single unique matches to the genome, while a small number of the MM probes (typically less 
than 1%) could be incorporated into the analysis.  In addition, we have examined the mapping of PM and MM probes to five
different human genome projects, resulting in approximately a 70% overlap of uniquely mapping PM probes, and a subset of
161 uniquely mapping MM probes commonly found in all five projects, 24 of which are found within annotated exonic regions. 
These results suggest that individual variation in transcriptome regions provides an additional complexity to microarray
data analysis. Given these results, we conclude that the development of custom chip definition files (CDFs) should
include MM probe sequences to provide the most effective means of transcriptome analysis of Affymetrix(r) GeneChip(r) 
arrays.

**Categories and Subject Descriptors**: J.3 [Life and Medical Sciences]: Biology and Genetics

**General Terms**: Algorithms, Measurement, Theory

**Keywords**: Bioinformatics, microarray, probe set, custom definition files.

## INTRODUCTION

Oligonucleotide-based microarray technologies provide a methodology whereby a researcher can indirectly measure the
expression level of an mRNA molecule being actively transcribed under a set of conditions by labeling a cDNA fragment
that hybridizes to a complementary probe sequence specific to a particular transcript.  Since their first use on
customized cDNA arrays [1] in the mid-1990s, they have been used as the de-facto standard for measuring global 
transcriptional changes under differing conditions.  While RNA-Seq [2] may eventually supplant microarrays as the 
method of choice, a large number of microarray experiments exist that have been deposited into publicly available
repositories such as NCBI’s Gene Expression Omnibus (GEO) [3] and EBI’s ArrayExpress [4].  As a case in point, GEO
contains 32,471 series as of 9/6/2012.  The majority of the entries in GEO were performed on arrays designed by
companies such as Affymetrix(r), Inc. (Santa Clara, CA), Agilent Technologies, Inc. (Santa Clara, CA), Illumina(r), Inc.
(San Diego, CA), and GE Healthcare Lifesciences (Piscataway, NJ), with nearly half of the series (16,181) being
performed on various Affymetrix® arrays. 

The design of Affymetrix(r) GeneChip(r) arrays in particular provides for probe sets consisting of 11, 16, or 20 distinct
25 base pair (BP) probes, with each probe having a corresponding perfect match (PM) and mismatch (MM) probe.  The PM
and MM differ by the exchange of the complementary base at the 13th position in the probe.  While MM probes were
originally designed to account for signal in the PM resulting from non-specific cross-hybridization, they are often
underutilized or completely ignored. Mismatch probes have been explored for use in long oligonucleotide arrays as well
[5], but their utilization is limited to the Affymetrix(r) platform. 

Affymetrix® provides a default GeneChip® analysis package known as the Micro Array Suite 5.0 (MAS 5.0) [6] that measures the signal intensity for a particular probe pair as: 

$$
	\begin{aligned}
	\text{signal} = TukeyBiweight ( \log (PM_{j} - MM^{*}_{j}) )
	\end{aligned}
$$

Where MM`*` is a modified version of MM that is never bigger than the intensity value of the PM.  The motivation behind
the modified mismatch intensity MM* is to report all probe-level intensities as positive values, and to remove the
influence of the minority of probes where the MM intensity value is significantly higher than the corresponding PM
intensity.  In addition to the intensity signal, MAS 5.0 also produces a detection p-value which flags a transcript as
"P" (present), "M" (marginal), or "A" (absent) based on the reliability of the probe set based on differences between
PM and MM intensities.

Known issues in the use of PM and MM probe intensities to generate a single probe set intensity values led to the
development of other approaches, including RMA [7] and GCRMA [8]  which completely ignore the MM probes. 

With the availability of individual probe and reference genome sequences, it is possible to re-map probes based on new
sources of genome annotations. This allows custom Chip Description Files (CDFs) wherein probes are grouped into novel
probe sets based on exon, transcript, and gene level annotation [9-22]. Most notable is the effort of the BrainArray
group [10] which updates custom CDFs for a large number of Affymetrix(r) GeneChips(r) by creating probe sets based on
annotated features such as Entrez Gene [23], Ensembl transcript, Ensembl gene, and RefSeq Gene [24].  Using custom CDFs
has been shown to impact the reliability of expression analysis [10, 20-22, 25]. However, to the authors’ knowledge,
only the PM probe sequences are used when generating custom CDFs.

```{r incTable1, include=FALSE}
tableCount <- incCount(tableCount, "t.genomeAssemblies")
```

Based on the observation that a small, yet significant number of PM-MM probe pairs exist where the MM intensity is
significantly increased over the PM intensity, our initial inclination was that these differences in intensities were
not due to cross-hybridization or rogue probes alone.  Therefore, keeping in mind that Affymetrix(r) probes have been
designed according to continually evolving genome assemblies, we proceeded to analyze PM and MM probes across eight
commonly studied species (`r I(pasteLabel("Table", tableCount, "t.genomeAssemblies"))`) by looking at PM and MM probes that uniquely map to the respective genome.

In addition to changing functional annotations, one potential problem area for microarray probe design is the presence
of single nucleotide polymorphisms (SNPs) within a population.  As the probes are designed using a reference genome or
transcriptome, a “one-size fits all” approach has been taken for the probes on a particular array.  However, SNPs are
known to occur relatively frequently throughout the genome, with build 137 of dbSNP [26] containing over 53.5 million
reference SNPs for the human genome. We have previously studied the effects of SNPs on Affymetrix(r) GeneChips(r) [27]
showing that a large number of SNPs lie in the areas where microarray probes have been designed.  This has been taken
into account in the BrainArray’s custom CDF files which incorporate SNP information.  To study the effects that
individual variation can play in microarray analysis, we looked at the mappings of PM and MM probes within five
distinct publicly available assemblies of human genomes.

## METHODS

```{r addTable2, include=FALSE}
tableCount <- incCount(tableCount, "t.affyGenechips")
tableCount <- incCount(tableCount, "t.humanAssemblies")
```

### Mapping of PM and MM Probes

Chromosomal-based genome assemblies were downloaded from the UCSC Goldenpath Genomes ftp server using an anonymous
login (ftp://hgdownload.cse.ucsc.edu/goldenPath/) [28] for eight commonly studied species, including *C. elegans*
(roundworm), *D. melanogaster* (fruit fly), *S. cerevisiae* (baker’s yeast), *X. tropicalis* (western clawed frog), *D. rerio*
(zebrafish), *M. musculus* (house mouse), *R. norvegicus* (brown Norway rat), and *H. sapiens* (human) (`r I(pasteLabel("Table", tableCount, "t.genomeAssemblies"))`).  Genome
indices were created using `bowtie-build` version 0.12.8 [29] with the default parameters.  Perfect match (PM) probe
sequences for Affymetrix(r) GeneChips(r) were obtained from Bioconductor (v 2.10) probe packages, which are constructed
from data available in NetAffx (`r I(pasteLabel("Table", tableCount, "t.affyGenechips"))`) with each new Bioconductor release. Mismatch (MM) probe sequences were
constructed by replacing the 13th base in the supplied PM probe sequence with the complementary base. PM and MM probes
were aligned to the indexed genomes using `bowtie` version 0.12.8 [29] with the parameters –v 0 and –a which used
together will report all valid probes matching with 100% identity.

<a id="t.genomeAssemblies"></a>

**`r I(pasteLabel("Table", tableCount, "t.genomeAssemblies", F))`. Genome assemblies used**

Organism | Reference Assembly | Build Date
---------|--------------------|-----------
*Caenorhabditis elegans* | ce6 | May 2008
*Drosophila melanogaster* | dm3 | Apr. 2006
*Saccaromyces cerevisiae* | sc3 | Apr. 2011
*Xenopus tropicalis* | xt3 | Nov. 2009
*Danio rerio* | dr6 | Dec. 2008
*Mus musculus* | mm10 | Dec. 2011
*Rattus norvegicus* | rn4 | Nov. 2004
*Homo sapiens* | hg19 | Feb. 2009

<a id="t.affyGenechips"></a>

**`r I(pasteLabel("Table", tableCount, "t.affyGenechips", F))`. Affymetrix(r) GeneChips(r) used**

Organism | GeneChip(r) Name
---------|-----------------
*Caenorhabditis elegans* | *C. elegans* Genome
*Drosophila melanogaster* | *Drosophila* Genome 2.0
*Saccaromyces cerevisiae* | Yeast Genome 2.0
*Xenopus tropicalis* | *Xenopus tropicalis* Genome
*Danio rerio* | Zebrafish Genome
*Mus musculus* | Mouse Genome 430 2.0
*Rattus norvegicus* | Rat Genome 230 2.0
*Homo sapiens* | Human Genome U133 Plus 2.0


###	Generation of Exons and Overlap

Exon regions were obtained from the UCSC genome browser as `BED` files with an entry for each exon. `mergeBed` from the
`bedTools` suite was used to merge overlapping exons from multiple transcripts into single contiguous exons. These
merged exons were used when defining overlaps of probe alignments with an exon. Probe and exon overlaps were defined
as any type of overlap with at least 23 bases overlapping on the same strand. Overlaps were determined using `Genomic Ranges` version 1.8.7 [31].

###	DNA Microarray Data

For each GeneChip(r), CEL files were downloaded from GEO for 20 random samples (with the exception of *S. cerevisiae* (12)
and *X. tropicalis* (4), the GSMs are listed in [gsmFiles.txt](https://raw.github.com/rmflight/affyMM/master/gsmFiles.txt)). Probe intensities were background
corrected using the MAS background correction method implemented in Bioconductor. Depending on the application,
intensities were log (base 2), square root transformed, or used as is. 

###	Negative PM-MM Set

A PM-MM set of probes was considered to be negative if nine (two for *S. cerevisiae* and six for *X. tropicalis*) or more
samples had a negative value for the difference in the PM-MM intensities. For examination of intensity distribution,
any PM-MM pair with a negative different greater than 1000 in one or more samples was considered and examined. 

###	Probe Correlations

For each MM probe that uniquely overlapped one merged exon (designated as a true-match MM, (TM^(mm))), the correlation
with all other MM probes in the probe set (mm) and the correlation with all other TM probes that also mapped uniquely
to the same exon (if there were three or more other probes also mapped to the exon) was calculated (tm). 

###	Human Variation

To gain an understanding of individual variation and the unique mapping of microarray probes, five whole genome
assemblies were downloaded for the human genome [32-36] (`r I(pasteLabel("Table", tableCount, "t.humanAssemblies"))`).   Probes from the HGU133APlus2.0 Affymetrix®
GeneChip® were aligned to each of these genomes using the methods previously described for mapping PM and MM probes.

<a id="t.humanAssemblies"></a>

**`r I(pasteLabel("Table", tableCount, "t.humanAssemblies", F))`. Whole human genome sequencing projects**

Name | Abbr. | Assembly Identifier | Bioproject | Number | Race
-----|-------|---------------------|------------|--------|-----
GRCh37 | Hg19 | 420368 | 31257 | Mixed
Hs`_`Celera`_`WGSA | Celera | 281338 | 1431 | Mixed^1
HuRefPrime | JCVI | 281188 | 19621 | Caucasian
BGIAF | BGI | 165398 | 42201 | African
HsapALLPATHS1 | HSAP1 | 238948 | 59877 | Caucasian

^(1)Celera assembly consists of one African-American, one Asian-Chinese, one Hispanic-Mexican, and two Caucasians.

## RESULTS

### Probes Matching Genomic Locations 

```{r probeMatches, echo=FALSE}
cntTable <- read.table("alignmentCounts.txt", sep="\t", header=T)
tableCount <- incCount(tableCount, "t.probeMatches")
cntTable$Organism <- paste("*", cntTable$Organism, "*", sep="")
cntTable <- cntTable[, c("Organism", "nPair", "map.PM", "map.MM", "singPM", "singMM")]
names(cntTable) <- c("Organism", "Number of Probe Pairs", "PM Mapped to Reference",
										 "MM Mapped to Reference", "PM Unique", "MM Unique")
tableOut <- tableCat(cntTable)
```

Given the PM probe sequences and the inferred MM sequences, individual probes were mapped to the corresponding genome
assembly as outlined in Methods.  The percentage of perfect match probes mapping to the genome ranged from a low of
80% (*X. tropicalis*) to a high of 95% (*D. melanogaster*), with the exception of *S. cerevisiae* (`r I(pasteLabel("Table", tableCount, "t.probeMatches", T))`).  It must be
noted that the lower percentage of *S. cerevisiae* matches (53%) is expected, as the Yeast Genome 2.0 GeneChip(r) contains
probes for two yeast species, *S. cerevisiae* and *S. pombe*. 

<a id="t.probeMatches"></a>

**`r I(pasteLabel("Table", tableCount, "t.probeMatches", F))`. GeneChip(r) probes mapping to reference genomes**

```{r insertProbeMatches, results='asis', echo=FALSE}
cat(tableOut, sep="\n")
```

Affymetrix(r) probesets are given suffix definitions depending upon the uniqueness of the exemplar sequence used to
design a probe set.  A designation of `_`at indicates the probe set perfectly matches a single transcript; `_`a`_`at
probe sets only perfectly match transcripts of the same gene; `_`s`_`at perfectly match multiple transcripts for the same
gene family; and `_`x`_`at indicates the probe set is identical or highly similar to other genes.  One of the
difficulties with these designations is that it relies upon a set of annotations at a particular point in time.  

Analysis of the probes that map to the genome (`r I(pasteLabel("Table", tableCount, "t.probeMatches"))`) indicates that 82% to 98% of the mapped probes map uniquely to
a single genomic location.  The fact that a number of probes map to multiple locations is not to be unexpected due to
the restrictions placed on probe set design.  However, it is expected that those probes mapping to multiple locations
would not be from the `_`at class of probes.  

```{r incFigure, include=FALSE}
figCount <- incCount(figCount, "f.allPlots")
tableCount <- incCount(tableCount, "t.probeType")
```

```{r probeTypeAnalysis, echo=FALSE}
translationTable <- matrix(ncol=3, byrow=T, data=c("ce6", "Ce", "celegans",
																									 "dm3", "Dm", "drosophila2",
																									 "dr6", "Dr", "zebrafish",
																									 "hg19", "Hs", "hsa",
																									 "mm10", "Mm", "mus",
																									 "rn4", "Rn", "rat",
																									 "sc3", "Sc", "yeast2",
																									 "xt3", "Xt", "xtropicalis"))
colnames(translationTable) <- c("short", "Organism", "array")

multFiles <- file.path("seqData", paste("multMatch_", translationTable[,"short"], ".txt", sep=""))
nOrg <- nrow(translationTable)
typeCounts <- matrix("", nrow=nOrg, ncol=5)
colnames(typeCounts) <- c("_x_at", "_s_at", "_a_at", "_at", "control")
rownames(typeCounts) <- translationTable[,"Organism"]
#probeType <- c("_x_at", "_s_at", "_at", "_a_at", "affx")
for (iOrg in 1:nOrg){
	multProbes <- scan(file=multFiles[iOrg], what=character())
	pClass <- probeClass(multProbes)
	multProbes <- multProbes[pClass == "pm"]
	totMult <- length(multProbes)
	
	cntrl <- grep("affx", multProbes, ignore.case=T)
	n_cntrl <- length(cntrl)
	p_cntrl <- format(n_cntrl / totMult * 100, digits=2)
	multProbes <- multProbes[-1*cntrl]
	
	x_at <- grep("_x_at", multProbes)
	n_xat <- length(x_at)
	p_xat <- format(n_xat / totMult * 100, digits=2)
	s_at <- grep("_s_at", multProbes)
	n_sat <- length(s_at)
	p_sat <- format(n_sat / totMult * 100, digits=2)
	a_at <- grep("_a_at", multProbes)
	n_aat <- length(a_at)
	p_aat <- format(n_aat / totMult * 100, digits=2)
	
	oProbes <- multProbes[-1*c(x_at, s_at, a_at)]
	o_at <- grep("_at", oProbes)
	n_oat <- length(o_at)
	p_oat <- format(n_oat / totMult * 100, digits=2)
	otherProbes <- length(oProbes[-1*o_at])
	pOther <- format(otherProbes / totMult * 100, digits=2)
	typeCounts[iOrg, "_x_at"] <- paste(n_xat, " (", p_xat, "%)", collapse="", sep="") 
	typeCounts[iOrg, "_s_at"] <- paste(n_sat, " (", p_sat, "%)", collapse="", sep="") 
	typeCounts[iOrg, "_a_at"] <- paste(n_aat, " (", p_aat, "%)", collapse="", sep="") 
	typeCounts[iOrg, "_at"] <- paste(n_oat, " (", p_oat, "%)", collapse="", sep="") 
	#typeCounts[iOrg, "other"] <- paste(otherProbes, " (", pOther, "%)", collapse="", sep="") 
	typeCounts[iOrg, "control"] <- paste(n_cntrl, " (", p_cntrl, "%)", collapse="", sep="")
}

typeCounts <- cbind(translationTable[,"Organism"], typeCounts)
typeCounts <- as.data.frame(typeCounts)
names(typeCounts)[1] <- "Organism"
typeCounts$Organism <- paste("*", typeCounts$Organism, "*", sep="")
typeTable <- tableCat(typeCounts)
# note: Eric gets 29 "other" probes for Ce here, and I think it is because he is counting control probes as other
```


To determine the reliability of these probes with the fluctuation of unknown transcripts, those probes that map with
100% identity to two or more locations in the genome were considered.  While these probes typically represent less
than 10% of the total number of probes for a given GeneChip(r), their classification could be important in detecting
cross hybridization.  One might expect that the greatest percentage of these would be within the `_`x`_`at and
`_`s`_`at classes.  However, as `r I(pasteLabel("Table", tableCount, "t.probeType", T))` shows, the larger genomes actually contain the greatest percentage in the
`_`at and `_`a`_`at classes, with anywhere from 18% (S. cerevisiae) to 91% (R. norvegicus) of the probes matching
multiple locations belonging to the `_`at class.  In addition, a small number of MM probes map to the genome as well. 
To better understand the effects this small set of MM probes might have on gene expression, we further reduced this to
a smaller subset where the mapping was within exon regions.  For these probes, we analyzed their signal intensities
from random samples compared to the overall distribution of PM and MM intensities, and the distribution of PM and MM
intensities within the corresponding exonic sequences (`r I(pasteLabel("Figure", figCount, "f.allPlots", T))`).  As these plots indicate, MM probes mapping within
exonic regions closely follow the expression density of PM probes mapping within exonic regions, and are significantly 
shifted from the overall expression profiles of MM probes.  These results suggest that while the number of these
probes is small, they offer significant information that should not be ignored, and furthermore, can confound analyses
where MM data is incorporated.

<a id="t.probeType"></a>

**`r I(pasteLabel("Table", tableCount, "t.probeType", F))`. Probe set classification of probes perfectly matching multiple genomic locations.** 

```{r insertTypeTable, results='asis', echo=FALSE}
cat(typeTable, sep="\n")
```

<a id="f.allPlots"></a>

```{r allPlots, echo=FALSE, fig.width=7, fig.height=4, dpi=300}
theme_set(theme_bw())
allData <- allData[!(names(allData) %in% "mm9")]

plotData <- adply(names(allData), 1, function(x){
	tmpInt <- allData[[x]]$allPlot$plotInt
	tmpInt$organism <- x
	return(tmpInt)
})

g <- ggplot(plotData, aes(x=probeInt, colour=probeType, linetype=probeType)) + geom_density() + facet_wrap(~organism, nrow=2) + labs(x="log2 probe intensity")
print(g)
```

**`r I(pasteLabel("Figure", figCount, "f.allPlots", F))`. Density profile of probe intensities (log2).**  PM: perfect match, MM: mismatch.  mm.all: background MM intensities; pm.all: background PM intensities; mm.exon: intensities of MM probes in exonic regions; pm.exon: intensities of PM probes in exonic regions.

```{r setPMMMFigs, include=FALSE}
figCount <- incCount(figCount, "f.pmmmDensityPlot")
figCount <- incCount(figCount, "f.pmmmIntensityPlot")
figCount <- incCount(figCount, "f.tmIntensityPlot")
figCount <- incCount(figCount, "f.correlationBox")
```

As some of these MM probes may bind to transcripts, we further considered those MM probes that uniquely mapped to
exons (irrespective of whether the corresponding PM probe mapped zero, one or multiple times to exons or the full
genome), examining the differences in signal intensity between the MM and its associated PM.  In many cases there is a
significant negative difference in the expression level of the PM-MM pair (an example is shown in Figures `r I(pasteLabel("", figCount, "f.pmmmDensityPlot", T))` and `r I(pasteLabel("", figCount, "f.pmmmIntensityPlot", T))`). 

<a id="f.pmmmDensityPlot"></a>

```{r pmmmDensity, echo=FALSE}
# need to get the experiment and probe set ID used for this figure and set it as a variable we can use!
mmIndx <- 12
negIndx <- 2

tmpDat <- allData$dr6$pmDifDat[[mmIndx]]
mmName <- tmpDat$invProbes[negIndx]
setName <- probeSetName(mmName)
expName <- tmpDat$probeData$gsmID[negIndx]

denPlot <- tmpDat$plotDat[[negIndx]]$denPlot

denPlot <- denPlot + ggtitle("")
print(denPlot)
```

**`r I(pasteLabel("Figure", figCount, "f.pmmmDensityPlot", F))`. Density of MM and PM probe set intensities not including the PM-MM pair that had a large negative difference.** Intensities from zebrafish probeset `r I(setName)`, in GEO sample
`r I(expName)`.

<a id="f.pmmmIntensityPlot"></a>

```{r pmmmIntensity, echo=FALSE}
probeSetPlot <- tmpDat$plotDat[[negIndx]]$psImage
probeSetPlot <- probeSetPlot + ggtitle("") + theme_grey()
print(probeSetPlot)
```

**`r I(pasteLabel("Figure", figCount, "f.pmmmIntensityPlot", F))`. Square root transformed intensities for each PM-MM pair.** The negative difference pair is at the extreme right end of the figure. Intensities from zebrafish probeset `r I(setName)`, in GEO sample `r I(expName)`.

If these MM probes are grouped instead with the other probes within the transcriptional region for which they uniquely
match  (we have renamed these probes as “true match” (TM) probes since they truly match the region in the genome),
there is a much better association between the probe intensities, as shown in `r I(pasteLabel("Figure", figCount, "f.tmIntensityPlot"))`.

<a id="f.tmIntensity"></a>

```{r tmIntensity, echo=FALSE}
intenPlot <- tmpDat$plotDat[[negIndx]]$exonImage
intenPlot <- intenPlot + ggtitle("")
print(intenPlot)
```

**`r I(pasteLabel("Figure", figCount, "f.tmIntensityPlot", F))`. Plot of probe set intensities for zebrafish where the TM probes overlap with the same TMmm probe in `r I(pasteLabel("Figure", figCount, "f.pmmmIntensityPlot", T))`** Intensities from zebrafish probeset `r I(setName)`, in GEO sample `r I(expName)`.

Further analysis was performed to test the correlation of the TM probes with the expression levels of both the
annotated probe group MM probes and with the TM-mapped transcript probes (`r I(pasteLabel("Figure", figCount, "f.correlationBox"))`).  The box plot in `r I(pasteLabel("Figure", figCount, "f.correlationBox"))` clearly
indicates that the TM intensities more closely correlate with those from the group based on mapping to the same exon.

<a id="f.correlationBox"></a>

```{r correlationBoxPlot, echo=FALSE, warning=FALSE}
allData <- allData[!(names(allData) %in% c("mm9", "sc3"))]
useName <- names(allData)

calcCor <- function(inN){
	useCor <- allData[[inN]]$mmCor
	mmProbes <- names(useCor)

	nCor <- length(useCor)
	
	mmCor <- data.frame(x=0, y=0, class="", orgName=inN, mmName="")
	tmCor <- data.frame(x=0, y=0, class="", orgName=inN, mmName="")
	
	for (iCor in 1:nCor){
		mmT <- as.vector(useCor[[iCor]]$mmCor)
		mmT <- data.frame(x=as.character(iCor), y=mmT, class="mm", orgName=inN, mmName=mmProbes[iCor])
		mmCor <- rbind(mmCor, mmT)
		
		tmT <- as.vector(useCor[[iCor]]$tmCor)
		if (!(is.na(tmT))){
			tmT <- data.frame(x=as.character(iCor), y=tmT, class="tm", orgName=inN, mmName=mmProbes[iCor])
			tmCor <- rbind(tmCor, tmT)
		}
	}
	
	mmCor <- mmCor[seq(2, nrow(mmCor)),]
	tmCor <- tmCor[seq(2, nrow(tmCor)),]
	
	allCor <- rbind(mmCor, tmCor)
	hasSame <- unique(mmCor$x[(mmCor$x %in% tmCor$x)])
	allCorSame <- allCor[(allCor$x %in% hasSame),]
	return(list(allCor=allCor, allCorSame=allCorSame))
}
corDat <- lapply(useName, calcCor)
names(corDat) <- useName

initCor <- corDat[[1]]$allCorSame
invisible(sapply(corDat[2:length(corDat)], function(x){
	initCor <<- rbind(initCor, x$allCorSame)
}))

initCor$orgName <- as.character(initCor$orgName)
initCor$orgName[(initCor$orgName == "xtropicalis")] <- "xt3"
initCor$class.orgName <- paste(initCor$orgName, initCor$class, sep=":")

ggplot(initCor, aes(x=orgName, y=y, colour=class)) + geom_boxplot() + theme_bw() + labs(x="organism", y="correlation")
```

**`r I(pasteLabel("Figure", figCount, "f.correlationBox", F))`. Box plot of correlations of the TMmm with MM intensities of annotated probe set (red, MM) and TM intensities of custom probe sets based on shared mapping to exons (blue, TM)**

```{r incCorrTable, echo=FALSE}
tableCount <- incCount(tableCount, "t.corrTable")
tableCount <- incCount(tableCount, "t.humanMapping")
```

To determine if the observed difference in the correlations from `r I(pasteLabel("Figure", figCount, "f.correlationBox"))` is due to measurement of different mRNA
entities, we considered the MM probe both within its annotated location as well as within the new mapped location
(TMmm). The intensity of the MM probe was compared with the intensity of its corresponding neighbor probes (MM probes
for the annotated location; PM probes for the new mapped location).  An average correlation value between intensities
was calculated on a per-exon probe basis.  The resulting correlations are summarized in `r I(pasteLabel("Table", tableCount, "t.corrTable"))`.

```{r addCorrTable, echo=FALSE, include=FALSE}
useCor <- corDat$hg19$allCorSame
mmProbes <- unique(useCor$mmName)

tmV <- useCor$class == "tm"
mmV <- useCor$class == "mm"
tmCor <- sapply(mmProbes, function(x){
	mean(useCor$y[(tmV & (useCor$mmName %in% x))])
})

mmCor <- sapply(mmProbes, function(x){
	mean(useCor$y[(mmV & (useCor$mmName %in% x))])
})

corInfo <- data.frame(tmCor=tmCor, mmCor=mmCor, mmProbe=mmProbes)
rownames(corInfo) <- NULL

# what are the annotated refseq for the probeset?
corInfo$probeSet <- sapply(as.character(corInfo$mmProbe), probeSetName)
annotRefSeq <- mget(corInfo$probeSet, hgu133plus2REFSEQ, ifnotfound=NA)
names(annotRefSeq) <- corInfo$mmProbe
annotRefSeq <- lapply(annotRefSeq, function(x){
	nmVal <- grep("NM", x, value=T)
	nrVal <- grep("NR", x, value=T)
	c(nmVal, nrVal)
}) # get the NMs back


probeAlign <- allData$hg19$probeAlign
probeAlign <- probeAlign[(elementMetadata(probeAlign)$probeNames %in% corInfo$mmProbe)]
exonIntervals <- allData$hg19$exonIntervals
geneExons <- allData$hg19$geneExons

probe2Exons <- as.matrix(findOverlaps(probeAlign, exonIntervals, minoverlap=23, type="any"))
ref2merge <- as.matrix(findOverlaps(geneExons, exonIntervals, type="within")) # the original exons should be "within" the merged exons
	
probeNames <- elementMetadata(probeAlign)$probeNames
refNames <- elementMetadata(geneExons)$refID	
probe2Ref <- lapply(seq(1, nrow(probe2Exons)), function(inRow){
	useExon <- probe2Exons[inRow,2]
	matchRef <- ref2merge[(ref2merge[,2] == useExon),1]
	refNames[matchRef]
})
names(probe2Ref) <- probeNames[probe2Exons[,1]]

invNames <- names(probe2Ref)
refComp <- lapply(invNames, function(inName){
	aRef <- annotRefSeq[[inName]]
	pRef <- probe2Ref[[inName]]
	aSpec <- aRef[!(aRef %in% pRef)]
	pSpec <- pRef[!(pRef %in% aRef)]
	return(list(aRef=aRef, aSpec=aSpec, pRef=pRef, pSpec=pSpec))
})

names(refComp) <- invNames
corInfo <- corInfo[order(corInfo$mmProbe),]
refComp <- refComp[order(names(refComp))]
corInfo$annotRef <- sapply(refComp, function(x){paste(x$aRef, collapse=", ")})
corInfo$exonRef <- sapply(refComp, function(x){paste(x$pRef, collapse=", ")})
corInfo <- corInfo[,c("tmCor", "mmCor", "probeSet", "mmProbe", "annotRef", "exonRef")]

tmpGene <- vector("list", length(refComp))
for (iGene in 1:length(refComp)){
	aRef <- refComp[[iGene]]$aRef
	if (length(aRef) != 0){
		tmpEnt <- unlist2(mget(aRef, org.Hs.egREFSEQ2EG, ifnotfound=NA))
		tmpSym <- unlist2(mget(tmpEnt, org.Hs.egSYMBOL, ifnotfound=NA))
		useEnt <- unique(names(tmpSym))
		for (iEnt in 1:length(useEnt)){
			tEnt <- useEnt[iEnt]
			names(tmpSym)[names(tmpSym) %in% tEnt] <- unique(names(tmpEnt)[tmpEnt %in% tEnt])
		}
		tmpGene[[iGene]]$aRef <- tmpSym
	}
	
	pRef <- refComp[[iGene]]$pRef
	if (length(pRef) != 0){
		tmpEnt <- unlist2(mget(pRef, org.Hs.egREFSEQ2EG, ifnotfound=NA))
		tmpSym <- unlist2(mget(tmpEnt, org.Hs.egSYMBOL, ifnotfound=NA))
		useEnt <- unique(names(tmpSym))
		for (iEnt in 1:length(useEnt)){
			tEnt <- useEnt[iEnt]
			names(tmpSym)[names(tmpSym) %in% tEnt] <- unique(names(tmpEnt)[tmpEnt %in% tEnt])
		}
		tmpGene[[iGene]]$pRef <- tmpSym
	}
	
}

names(tmpGene) <- names(refComp)
corInfo$aSym <- sapply(tmpGene, function(x){paste(unique(x$aRef), collapse=", ")})
corInfo$pSym <- sapply(tmpGene, function(x){paste(unique(x$pRef), collapse=", ")})
corOut <- corInfo
names(corOut) <- c("tm", "mm", "ProbeSet", "Probe ID", "Annotated RefSeq", "Exon RefSeq", "Annotated Symbol", "Exon Symbol")
corOut <- corOut[order(corOut$tm, decreasing=T),]
corOut2 <- tableCat(corOut)
```

<a id="t.corrTable"></a>

```{r insertCorTable, results='asis', echo=FALSE}
cat(corOut2, sep="\n")
```

What is interesting is that with few exceptions, the transcripts and genes being measured by the probe sets are the
same, implying that the TMmm probe aligns to the same gene as its complimentary PM probe. Upon further examination it
appears that many of the PM complements of the TMmm do not align to the genome at all (data not shown). One possibility
is that these probes are in regions that have seen changes in the reference sequence over the years, or that the
original sequencing of the ESTs used to design the probe sets was of poor quality. 

###	Effects of Individual Variation 

```{r humanMappingResults, echo=FALSE, include=FALSE}
hgVariationFiles <- c(Hg19="hg.txt", Celera="celera.txt", JCVI="jcvi.txt", BGI="bgi.txt", HSAP1="hsapl.txt")

hgAlignRes <- alply(hgVariationFiles, 1, function(x){
	scan(file=file.path("hgs_uniquelyMapped_probes", x), what=character())
})

probeCounts <- ldply(hgAlignRes, function(x){
	nTot <- length(x)
	tmpClass <- probeClass(x)
	nPM <- sum(tmpClass == "pm")
	nMM <- sum(tmpClass == "mm")
	return(c(nTot, nPM, nMM))
})

names(probeCounts) <- c("Assembly", "Total", "Perfect Match", "Mismatch")
probeCounts2 <- tableCat(probeCounts)
```

To gain an understanding of the effect of individual variation, the unique mapping of PM and MM probes to five
distinct human genome assemblies was analyzed (`r I(pasteLabel("Table", tableCount, "t.humanMapping"))`).  Four of the five projects have roughly the same number of
uniquely mapped PM probes (within 2% variation).  The fifth project (BGI) provides an exception to this trend.  While
there are a number of potential explanations for this (including sequence and assembly quality and coverage of the
sequencing), one potential feature to be considered is the fact that this sequencing project involves the sequencing
of an African individual, and it is the only project not to have a large component of the library consisting of
Caucasian individuals.

<a id="t.humanMapping"></a>

**`r I(pasteLabel("Table", tableCount, "t.humanMapping", F))`. Number of probes mapping uniquely to individual human genomes.**

```{r insertHumanMapping, results='asis', echo=FALSE}
cat(probeCounts2, sep="\n")
```

```{r incVennFigure, include=FALSE}
figCount <- incCount(figCount, "f.vennHumanVariation")

# get the actual counts of individual and intersections
qVennArgs <- list()

count <- 1
for (iVal in 1:length(hgAlignRes)){
	useName <- paste("area", iVal, sep="", collapse="")
	qVennArgs[[iVal]] <- length(hgAlignRes[[iVal]])
	names(qVennArgs)[iVal] <- useName
	count <- count + 1
}

# get the intersection of each 2, 3, 4, and 5 way combination
allIntersection <- function(listIndices){
	nIntersect <- length(listIndices)
	foundIntersection <- hgAlignRes[[listIndices[1]]]
	for (iI in 2:nIntersect){
		foundIntersection <- intersect(foundIntersection, hgAlignRes[[listIndices[iI]]])
	}
	return(foundIntersection)
}

getIntersection <- combn(seq(1,5), 2, simplify=F)
getIntersection <- c(getIntersection, combn(seq(1,5), 3, simplify=F))
getIntersection <- c(getIntersection, combn(seq(1,5), 4, simplify=F))
getIntersection <- c(getIntersection, combn(seq(1,5), 5, simplify=F))

whichIntersect <- lapply(getIntersection, allIntersection)
sumIntersect <- lapply(whichIntersect, length)

intersectLabels <- sapply(getIntersection, function(x){
	paste("n", paste(x, sep="", collapse=""), sep="", collapse="")
})

names(sumIntersect) <- intersectLabels
names(whichIntersect) <- intersectLabels

qVennArgs <- c(qVennArgs, sumIntersect)
qVennArgs$category <- probeCounts$Assembly
qVennArgs$fill <- c("dodgerblue", "goldenrod1", "darkorange1", "seagreen3", "orchid3")
qVennArgs$col <- "none"
#qVennArgs$cat.col <- c("dodgerblue", "goldenrod1", "darkorange1", "seagreen3", "orchid3")
qVennArgs$cat.cex <- 2
qVennArgs$margin <- 0.05
qVennArgs$cex <- c(1.5, 1.5, 1.5, 1.5, 1.5, 1, 0.8, 1, 0.8, 1, 0.8, 1, 0.8, 1, 0.8, 1, 0.55, 1, 0.55, 1, 0.55, 1, 0.55, 1, 0.55, 1, 1, 1, 1, 1, 1.5)
qVennArgs$ind <- TRUE
```

```{r intersectStats, include=FALSE}
all5Intersect <- whichIntersect$n12345
all5Class <- probeClass(all5Intersect)

data(hgu133plus2probe)
nPM <- nrow(hgu133plus2probe)

goodPM <- sum(all5Class == "pm")
percPM <- goodPM / nPM * 100

goodMM <- all5Intersect[all5Class == "mm"]

# which mm fall within exonic regions? This is not the best because it only
# uses the exons that overlapped singleAnnot originally
probeAlign <- allData$hg19$probeAlign
probeAlign <- probeAlign[(elementMetadata(probeAlign)$probeNames %in% goodMM)]
isExon <- subsetByOverlaps(probeAlign, exonIntervals, minoverlap=23)

isExonNames <- elementMetadata(isExon)$probeNames
tmVmm <- corOut[(corOut$'Probe ID' %in% isExonNames), c('tm','mm')]
```

While there is a large agreement for the number of probes, we also checked if the probes represented were consistent
among all of the projects.  A Venn diagram depicting the number of overlapping perfectly matching probes is given in
`r I(pasteLabel("Figure", figCount, "f.vennHumanVariation"))`.  As can be seen from this figure, a total of `r I(length(all5Intersect))` probes uniquely map for all five assemblies.  Of these,
`r I(goodPM)` are perfect match probes, indicating that `r I(format(percPM, digits=2))`% of the HGU133APlus2.0 perfect match probes are reliable in terms
of their mapping to the genome for these assemblies.  One of the interesting results is that there are `r I(length(goodMM))` shared
perfectly matching mismatch probes.  Of these, `r I(length(isExon))` fall within RefSeq annotated exonic regions (results not shown),
with `r I(sum(tmVmm["tm"] >= tmVmm["mm"]))` of the `r I(length(isExon))` showing higher correlation in the TMmm assignments calculated in `r I(pasteLabel("Table", tableCount, "t.corrTable"))`.

<a id="f.vennHumanVariation"></a>

```{r insertVenn, echo=FALSE}
venPlot <- do.call("draw.quintuple.venn", qVennArgs) 
```

**`r I(pasteLabel("Figure", figCount, "f.vennHumanVariation", F))`. Venn diagram of overlapping perfectly matching Affymetrix(r) HGU133A Plus2 probes to each of the five human genome assemblies.**

## Conclusion

MM probes are theoretically designed to capture background and non-specific binding. Alignment of the MM probes to the
genome shows that in a very small percentage of cases, MM probes align uniquely to the genome in transcribed regions.
Signal from these probes should be useful for quantifying true transcriptional events rather than for PM signal
adjustment.

In addition, current custom CDF generation workflows ignore the MM probes during the probe alignment process. Given
that some MM probes align to reference genomes, they should be considered for inclusion when creating custom CDFs. The
utility of the probes may be limited due to variation among individuals. 

## ACKNOWLEDGEMENTS

This work was partially funded by National Institutes of Health (NIH) grant 8P20GM103436-12.  Its contents are solely
the responsibility of the authors and do not represent the official views of NIH or the National Institute of General
Medical Sciences.

## REFERENCES

[1] Schena, M., Shalon, D., Davis, R. W. and Brown, P. O. Quantitative monitoring of gene expression patterns with a complementary DNA microarray. Science, 270, 5235 (Oct 20 1995), 467-470.

[2] Nagalakshmi, U., Wang, Z., Waern, K., Shou, C., Raha, D., Gerstein, M. and Snyder, M. The transcriptional landscape of the yeast genome defined by RNA sequencing. Science, 320, 5881 (Jun 6 2008), 1344-1349.

[3] Barrett, T., Troup, D. B., Wilhite, S. E., Ledoux, P., Evangelista, C., Kim, I. F., Tomashevsky, M., Marshall, K. A., Phillippy, K. H., Sherman, P. M., Muertter, R. N., Holko, M., Ayanbule, O., Yefanov, A. and Soboleva, A. NCBI GEO: archive for functional genomics data sets--10 years on. Nucleic acids research, 39, Database issue (Jan 2011), D1005-1010.

[4] Parkinson, H., Sarkans, U., Kolesnikov, N., Abeygunawardena, N., Burdett, T., Dylag, M., Emam, I., Farne, A., Hastings, E., Holloway, E., Kurbatova, N., Lukk, M., Malone, J., Mani, R., Pilicheva, E., Rustici, G., Sharma, A., Williams, E., Adamusiak, T., Brandizi, M., Sklyar, N. and Brazma, A. ArrayExpress update--an archive of microarray and high-throughput sequencing-based functional genomics experiments. Nucleic acids research, 39, Database issue (Jan 2011), D1002-1004.

[5] Deng, Y., He, Z., Van Nostrand, J. D. and Zhou, J. Design and analysis of mismatch probes for long oligonucleotide microarrays. BMC genomics, 92008), 491.

[6] Hubbell, E., Liu, W. M. and Mei, R. Robust estimators for expression analysis. Bioinformatics, 18, 12 (Dec 2002), 1585-1592.

[7] Irizarry, R. A., Bolstad, B. M., Collin, F., Cope, L. M., Hobbs, B. and Speed, T. P. Summaries of Affymetrix GeneChip probe level data. Nucleic acids research, 31, 4 (Feb 15 2003), e15.

[8] Wu, Z. and Irizarry, R. A. Stochastic models inspired by hybridization theory for short oligonucleotide arrays. Journal of computational biology : a journal of computational molecular cell biology, 12, 6 (Jul-Aug 2005), 882-893.

[9] Carter, S. L., Eklund, A. C., Mecham, B. H., Kohane, I. S. and Szallasi, Z. Redefinition of Affymetrix probe sets by sequence overlap with cDNA microarray probes reduces cross-platform inconsistencies in cancer-associated gene expression measurements. BMC bioinformatics, 62005), 107.

[10] Dai, M., Wang, P., Boyd, A. D., Kostov, G., Athey, B., Jones, E. G., Bunney, W. E., Myers, R. M., Speed, T. P., Akil, H., Watson, S. J. and Meng, F. Evolving gene/transcript definitions significantly alter the interpretation of GeneChip data. Nucleic acids research, 33, 20 2005), e175.

[11] de Leeuw, W. C., Rauwerda, H., Jonker, M. J. and Breit, T. M. Salvaging Affymetrix probes after probe-level re-annotation. BMC research notes, 12008), 66.

[12] Ferrari, F., Bortoluzzi, S., Coppe, A., Sirota, A., Safran, M., Shmoish, M., Ferrari, S., Lancet, D., Danieli, G. A. and Bicciato, S. Novel definition files for human GeneChips based on GeneAnnot. BMC bioinformatics, 82007), 446.

[13] Langer, W., Sohler, F., Leder, G., Beckmann, G., Seidel, H., Grone, J., Hummel, M. and Sommer, A. Exon array analysis using re-defined probe sets results in reliable identification of alternatively spliced genes in non-small cell lung cancer. BMC genomics, 112010), 676.

[14] Liu, H., Zeeberg, B. R., Qu, G., Koru, A. G., Ferrucci, A., Kahn, A., Ryan, M. C., Nuhanovic, A., Munson, P. J., Reinhold, W. C., Kane, D. W. and Weinstein, J. N. AffyProbeMiner: a web resource for computing or retrieving accurately redefined Affymetrix probe sets. Bioinformatics, 23, 18 (Sep 15 2007), 2385-2390.

[15] Lu, J., Lee, J. C., Salit, M. L. and Cam, M. C. Transcript-based redefinition of grouped oligonucleotide probe sets using AceView: high-resolution annotation for microarrays. BMC bioinformatics, 82007), 108.

[16] Moll, A. G., Lindenmeyer, M. T., Kretzler, M., Nelson, P. J., Zimmer, R. and Cohen, C. D. Transcript-specific expression profiles derived from sequence-based analysis of standard microarrays. PloS one, 4, 3 2009), e4702.

[17] Moreews, F., Rauffet, G., Dehais, P. and Klopp, C. SigReannot-mart: a query environment for expression microarray probe re-annotations. Database : the journal of biological databases and curation, 20112011), bar025.

[18] Neerincx, P. B., Rauwerda, H., Nie, H., Groenen, M. A., Breit, T. M. and Leunissen, J. A. OligoRAP - an Oligo Re-Annotation Pipeline to improve annotation and estimate target specificity. BMC proceedings, 3 Suppl 42009), S4.

[19] Risueno, A., Fontanillo, C., Dinger, M. E. and De Las Rivas, J. GATExplorer: genomic and transcriptomic explorer; mapping expression probes to gene loci, transcripts, exons and ncRNAs. BMC bioinformatics, 112010), 221.

[20] Sandberg, R. and Larsson, O. Improved precision and accuracy for microarrays using updated probe set definitions. BMC bioinformatics, 82007), 48.

[21] Yin, J., McLoughlin, S., Jeffery, I. B., Glaviano, A., Kennedy, B. and Higgins, D. G. Integrating multiple genome annotation databases improves the interpretation of microarray gene expression data. BMC genomics, 112010), 50.

[22] Yu, H., Wang, F., Tu, K., Xie, L., Li, Y. Y. and Li, Y. X. Transcript-level annotation of Affymetrix probesets improves the interpretation of gene expression data. BMC bioinformatics, 82007), 194.

[23] Maglott, D., Ostell, J., Pruitt, K. D. and Tatusova, T. Entrez Gene: gene-centered information at NCBI. Nucleic acids research, 39, Database issue (Jan 2011), D52-57.

[24] Pruitt, K. D., Tatusova, T., Brown, G. R. and Maglott, D. R. NCBI Reference Sequences (RefSeq): current status, new features and genome annotation policy. Nucleic acids research, 40, Database issue (Jan 2012), D130-135.

[25] Lu, X. and Zhang, X. The effect of GeneChip gene definitions on the microarray study of cancers. BioEssays : news and reviews in molecular, cellular and developmental biology, 28, 7 (Jul 2006), 739-746.

[26] Sayers, E. W., Barrett, T., Benson, D. A., Bolton, E., Bryant, S. H., Canese, K., Chetvernin, V., Church, D. M., Dicuccio, M., Federhen, S., Feolo, M., Fingerman, I. M., Geer, L. Y., Helmberg, W., Kapustin, Y., Krasnov, S., Landsman, D., Lipman, D. J., Lu, Z., Madden, T. L., Madej, T., Maglott, D. R., Marchler-Bauer, A., Miller, V., Karsch-Mizrachi, I., Ostell, J., Panchenko, A., Phan, L., Pruitt, K. D., Schuler, G. D., Sequeira, E., Sherry, S. T., Shumway, M., Sirotkin, K., Slotta, D., Souvorov, A., Starchenko, G., Tatusova, T. A., Wagner, L., Wang, Y., Wilbur, W. J., Yaschenko, E. and Ye, J. Database resources of the National Center for Biotechnology Information. Nucleic acids research, 40, Database issue (Jan 2012), D13-25.

[27] Rouchka, E. C., Phatak, A. W. and Singh, A. V. Effect of single nucleotide polymorphisms on Affymetrix match-mismatch probe pairs. Bioinformation, 2, 9 2008), 405-411.

[28] Dreszer, T. R., Karolchik, D., Zweig, A. S., Hinrichs, A. S., Raney, B. J., Kuhn, R. M., Meyer, L. R., Wong, M., Sloan, C. A., Rosenbloom, K. R., Roe, G., Rhead, B., Pohl, A., Malladi, V. S., Li, C. H., Learned, K., Kirkup, V., Hsu, F., Harte, R. A., Guruvadoo, L., Goldman, M., Giardine, B. M., Fujita, P. A., Diekhans, M., Cline, M. S., Clawson, H., Barber, G. P., Haussler, D. and James Kent, W. The UCSC Genome Browser database: extensions and updates 2011. Nucleic acids research, 40, Database issue (Jan 2012), D918-923.

[29] Langmead, B., Trapnell, C., Pop, M. and Salzberg, S. L. Ultrafast and memory-efficient alignment of short DNA sequences to the human genome. Genome biology, 10, 3 2009), R25.

[30] Liu, G., Loraine, A. E., Shigeta, R., Cline, M., Cheng, J., Valmeekam, V., Sun, S., Kulp, D. and Siani-Rose, M. A. NetAffx: Affymetrix probesets and annotations. Nucleic acids research, 31, 1 (Jan 1 2003), 82-86.

[31] Aboyoun, H., Pages, H. and Lawrence, M. GenomicRanges: Representation and manipulation of genomic intervals. . City.

[32] Gnerre, S., Maccallum, I., Przybylski, D., Ribeiro, F. J., Burton, J. N., Walker, B. J., Sharpe, T., Hall, G., Shea, T. P., Sykes, S., Berlin, A. M., Aird, D., Costello, M., Daza, R., Williams, L., Nicol, R., Gnirke, A., Nusbaum, C., Lander, E. S. and Jaffe, D. B. High-quality draft assemblies of mammalian genomes from massively parallel sequence data. Proceedings of the National Academy of Sciences of the United States of America, 108, 4 (Jan 25 2011), 1513-1518.

[33] Istrail, S., Sutton, G. G., Florea, L., Halpern, A. L., Mobarry, C. M., Lippert, R., Walenz, B., Shatkay, H., Dew, I., Miller, J. R., Flanigan, M. J., Edwards, N. J., Bolanos, R., Fasulo, D., Halldorsson, B. V., Hannenhalli, S., Turner, R., Yooseph, S., Lu, F., Nusskern, D. R., Shue, B. C., Zheng, X. H., Zhong, F., Delcher, A. L., Huson, D. H., Kravitz, S. A., Mouchard, L., Reinert, K., Remington, K. A., Clark, A. G., Waterman, M. S., Eichler, E. E., Adams, M. D., Hunkapiller, M. W., Myers, E. W. and Venter, J. C. Whole-genome shotgun assembly and comparison of human genome assemblies. Proceedings of the National Academy of Sciences of the United States of America, 101, 7 (Feb 17 2004), 1916-1921.

[34] Lander, E. S., Linton, L. M., Birren, B., Nusbaum, C., Zody, M. C., Baldwin, J., Devon, K., Dewar, K., Doyle, M., FitzHugh, W., Funke, R., Gage, D., Harris, K., Heaford, A., Howland, J., Kann, L., Lehoczky, J., LeVine, R., McEwan, P., McKernan, K., Meldrim, J., Mesirov, J. P., Miranda, C., Morris, W., Naylor, J., Raymond, C., Rosetti, M., Santos, R., Sheridan, A., Sougnez, C., Stange-Thomann, N., Stojanovic, N., Subramanian, A., Wyman, D., Rogers, J., Sulston, J., Ainscough, R., Beck, S., Bentley, D., Burton, J., Clee, C., Carter, N., Coulson, A., Deadman, R., Deloukas, P., Dunham, A., Dunham, I., Durbin, R., French, L., Grafham, D., Gregory, S., Hubbard, T., Humphray, S., Hunt, A., Jones, M., Lloyd, C., McMurray, A., Matthews, L., Mercer, S., Milne, S., Mullikin, J. C., Mungall, A., Plumb, R., Ross, M., Shownkeen, R., Sims, S., Waterston, R. H., Wilson, R. K., Hillier, L. W., McPherson, J. D., Marra, M. A., Mardis, E. R., Fulton, L. A., Chinwalla, A. T., Pepin, K. H., Gish, W. R., Chissoe, S. L., Wendl, M. C., Delehaunty, K. D., Miner, T. L., Delehaunty, A., Kramer, J. B., Cook, L. L., Fulton, R. S., Johnson, D. L., Minx, P. J., Clifton, S. W., Hawkins, T., Branscomb, E., Predki, P., Richardson, P., Wenning, S., Slezak, T., Doggett, N., Cheng, J. F., Olsen, A., Lucas, S., Elkin, C., Uberbacher, E., Frazier, M., Gibbs, R. A., Muzny, D. M., Scherer, S. E., Bouck, J. B., Sodergren, E. J., Worley, K. C., Rives, C. M., Gorrell, J. H., Metzker, M. L., Naylor, S. L., Kucherlapati, R. S., Nelson, D. L., Weinstock, G. M., Sakaki, Y., Fujiyama, A., Hattori, M., Yada, T., Toyoda, A., Itoh, T., Kawagoe, C., Watanabe, H., Totoki, Y., Taylor, T., Weissenbach, J., Heilig, R., Saurin, W., Artiguenave, F., Brottier, P., Bruls, T., Pelletier, E., Robert, C., Wincker, P., Smith, D. R., Doucette-Stamm, L., Rubenfield, M., Weinstock, K., Lee, H. M., Dubois, J., Rosenthal, A., Platzer, M., Nyakatura, G., Taudien, S., Rump, A., Yang, H., Yu, J., Wang, J., Huang, G., Gu, J., Hood, L., Rowen, L., Madan, A., Qin, S., Davis, R. W., Federspiel, N. A., Abola, A. P., Proctor, M. J., Myers, R. M., Schmutz, J., Dickson, M., Grimwood, J., Cox, D. R., Olson, M. V., Kaul, R., Shimizu, N., Kawasaki, K., Minoshima, S., Evans, G. A., Athanasiou, M., Schultz, R., Roe, B. A., Chen, F., Pan, H., Ramser, J., Lehrach, H., Reinhardt, R., McCombie, W. R., de la Bastide, M., Dedhia, N., Blocker, H., Hornischer, K., Nordsiek, G., Agarwala, R., Aravind, L., Bailey, J. A., Bateman, A., Batzoglou, S., Birney, E., Bork, P., Brown, D. G., Burge, C. B., Cerutti, L., Chen, H. C., Church, D., Clamp, M., Copley, R. R., Doerks, T., Eddy, S. R., Eichler, E. E., Furey, T. S., Galagan, J., Gilbert, J. G., Harmon, C., Hayashizaki, Y., Haussler, D., Hermjakob, H., Hokamp, K., Jang, W., Johnson, L. S., Jones, T. A., Kasif, S., Kaspryzk, A., Kennedy, S., Kent, W. J., Kitts, P., Koonin, E. V., Korf, I., Kulp, D., Lancet, D., Lowe, T. M., McLysaght, A., Mikkelsen, T., Moran, J. V., Mulder, N., Pollara, V. J., Ponting, C. P., Schuler, G., Schultz, J., Slater, G., Smit, A. F., Stupka, E., Szustakowski, J., Thierry-Mieg, D., Thierry-Mieg, J., Wagner, L., Wallis, J., Wheeler, R., Williams, A., Wolf, Y. I., Wolfe, K. H., Yang, S. P., Yeh, R. F., Collins, F., Guyer, M. S., Peterson, J., Felsenfeld, A., Wetterstrand, K. A., Patrinos, A., Morgan, M. J., de Jong, P., Catanese, J. J., Osoegawa, K., Shizuya, H., Choi, S. and Chen, Y. J. Initial sequencing and analysis of the human genome. Nature, 409, 6822 (Feb 15 2001), 860-921.

[35] Levy, S., Sutton, G., Ng, P. C., Feuk, L., Halpern, A. L., Walenz, B. P., Axelrod, N., Huang, J., Kirkness, E. F., Denisov, G., Lin, Y., MacDonald, J. R., Pang, A. W., Shago, M., Stockwell, T. B., Tsiamouri, A., Bafna, V., Bansal, V., Kravitz, S. A., Busam, D. A., Beeson, K. Y., McIntosh, T. C., Remington, K. A., Abril, J. F., Gill, J., Borman, J., Rogers, Y. H., Frazier, M. E., Scherer, S. W., Strausberg, R. L. and Venter, J. C. The diploid genome sequence of an individual human. PLoS biology, 5, 10 (Sep 4 2007), e254.

[36] Li, R., Li, Y., Zheng, H., Luo, R., Zhu, H., Li, Q., Qian, W., Ren, Y., Tian, G., Li, J., Zhou, G., Zhu, X., Wu, H., Qin, J., Jin, X., Li, D., Cao, H., Hu, X., Blanche, H., Cann, H., Zhang, X., Li, S., Bolund, L., Kristiansen, K., Yang, H. and Wang, J. Building the sequence map of the human pan-genome. Nature biotechnology, 28, 1 (Jan 2010), 57-63.

## RECREATION OF RESULTS

All of the necessary code to regenerate this work is hosted at https://github.com/rmflight/affymm. To recreate this html document, you need to run:

`knit2html("flightetal_draft.Rmd", "flightetal_draft.html")`

Alternatively use the `Knit HTML` button in `RStudio v0.97` or newer. This document was
generated using `r R.version.string`. The following information about packages used is also supplied:

```{r writeSessionInfo, comment=""}
sessionInfo()
```